tpie/pipelining/node.cpp

```diff
@@ -226,40 +226,42 @@ void node::add_forwarded_data(std::string key, node_token::id_t from_node) {
 	m_forwardedToHere[key] = from_node;
 }
 
-boost::any node::fetch_any(std::string key, const std::string & type_name) {
+node::maybeany_t node::fetch_maybe(std::string key) {
+	bits::node_map::ptr nodeMap = get_node_map()->find_authority();
+
 	auto it = m_forwardedToHere.find(key);
 	if (it == m_forwardedToHere.end()) {
-		std::stringstream ss;
-		ss << "Tried to fetch nonexistent key '" << key
-		   << "' of type " << type_name
-		   << " in " << get_name() << " of type " << typeid(*this).name();
-		throw invalid_argument_exception(ss.str());
+		// Try to lookup the key in the node map
+		return nodeMap->fetch_maybe(key);
 	}
 
 	node_token::id_t fetch_from_id = it->second;
-	bits::node_map::ptr nodeMap = get_node_map()->find_authority();
 	node *fetch_from = nodeMap->get(fetch_from_id);
 	if (!fetch_from) {
+		return maybeany_t();
+	}
+
+	return fetch_from->get_forwarded_data_maybe(key);
+}
```

Comment:
I think it is clearer when using optional and it also allows you to use `value_or`